### ISSUE TITLE
Use of implicit union return type causes  "'boolean' is not assignable to type 'false'"

### ISSUE THREAD

**Issue by jimkyndemeyer** [Microsoft/TypeScript#19212]
**TypeScript Version:**  2.6.0-dev.20171015

This relates to the render method of React components, specifically the use of `false` in the union return type:
```
render(): JSX.Element | null | false
```

I've create a stand-alone example that demonstrates the issue. See the code comments for instructions.

**Code**

```ts
declare class Base {
    render(): Date | false;
}

class TestImplicitReturnType extends Base {

    render() /* ': false' is required here as explicit return type to prevent a compile error */ {
       return false;
    }

}

class TestControlFlow extends Base {

    render() {
        if (new Date().getTime() % 2 === 0) {
            return new Date(); /* removing this return statement causes the false value below to be rejected  */ 
        }
        return false;
    }

} 

```

**Expected behavior:**
For TestImplicitReturn I should not be required to manually add `: false` as return type.

For TestControlFlow I should not be required to also return a Date to satisfy the type checker.

**Actual behavior:**

For TestImplicitReturn the compiler outputs:
```
Class 'TestImplicitReturnType' incorrectly extends base class 'Base'.
  Types of property 'render' are incompatible.
    Type '() => boolean' is not assignable to type '() => false | Date'.
      Type 'boolean' is not assignable to type 'false | Date'.
```

For TestImplicitReturnType the compiler outputs the following when the date return statement is removed:
```
Class 'TestControlFlow' incorrectly extends base class 'Base'.
  Types of property 'render' are incompatible.
    Type '() => boolean' is not assignable to type '() => false | Date'.
      Type 'boolean' is not assignable to type 'false | Date'.
```

**Comment by mhegazy** [comment]
The issue here is that a method in a derived class is not contextually typed by the type of the base's method. This has been discussed in by https://github.com/Microsoft/TypeScript/issues/1373  (and related to https://github.com/Microsoft/TypeScript/issues/3667, https://github.com/Microsoft/TypeScript/pull/6118 and https://github.com/Microsoft/TypeScript/pull/10610).

**Comment by mhegazy** [comment]
Duplicate of  https://github.com/Microsoft/TypeScript/issues/1373

**Comment by jimkyndemeyer** [comment]
Thanks for the detailed reply.

I still find it odd (counter-intuitive) that `false` produces a compilation error when `boolean` doesn't:

[TypeScript playground using boolean](https://www.typescriptlang.org/play/#src=declare%20class%20Base%20%7B%0D%0A%20%20%20%20render()%3A%20Date%20%7C%20boolean%3B%20%2F%2F%20use%20'boolean'%20instead%20of%20'false'%20here%20and%20then%20it%20compiles%0D%0A%7D%0D%0A%0D%0Aclass%20TestImplicitReturnType%20extends%20Base%20%7B%0D%0A%0D%0A%20%20%20%20render()%20%2F*%20'%3A%20false'%20is%20required%20here%20as%20explicit%20return%20type%20to%20prevent%20a%20compile%20error%20*%2F%20%7B%0D%0A%20%20%20%20%20%20%20return%20false%3B%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%7D%0D%0A%0D%0Aclass%20TestControlFlow%20extends%20Base%20%7B%0D%0A%0D%0A%20%20%20%20render()%20%7B%0D%0A%20%20%20%20%20%20%20%20if%20(new%20Date().getTime()%20%25%202%20%3D%3D%3D%200)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2Freturn%20new%20Date()%3B%20%2F*%20removing%20this%20return%20statement%20causes%20the%20false%20value%20below%20to%20be%20rejected%20%20*%2F%20%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20return%20false%3B%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%7D)


[TypeScript playground using false](https://www.typescriptlang.org/play/#src=declare%20class%20Base%20%7B%0D%0A%20%20%20%20render()%3A%20Date%20%7C%20false%3B%20%2F%2F%20use%20'boolean'%20instead%20of%20'false'%20here%20and%20then%20it%20compiles%0D%0A%7D%0D%0A%0D%0Aclass%20TestImplicitReturnType%20extends%20Base%20%7B%0D%0A%0D%0A%20%20%20%20render()%20%2F*%20'%3A%20false'%20is%20required%20here%20as%20explicit%20return%20type%20to%20prevent%20a%20compile%20error%20*%2F%20%7B%0D%0A%20%20%20%20%20%20%20return%20false%3B%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%7D%0D%0A%0D%0Aclass%20TestControlFlow%20extends%20Base%20%7B%0D%0A%0D%0A%20%20%20%20render()%20%7B%0D%0A%20%20%20%20%20%20%20%20if%20(new%20Date().getTime()%20%25%202%20%3D%3D%3D%200)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2Freturn%20new%20Date()%3B%20%2F*%20removing%20this%20return%20statement%20causes%20the%20false%20value%20below%20to%20be%20rejected%20%20*%2F%20%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20return%20false%3B%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%7D)

Does this narrow the scope of a potential fix? As it stands we're seeing the unwanted compilation error when using [@graphql](https://github.com/apollographql/react-apollo/blob/3784abc6c55422e77c0120a4f3e11f532cc8900f/src/graphql.tsx#L68-L75).

**Comment by RyanCavanaugh** [comment]
As the linked issue mentions, we don't currently use information about the base class to inform the type of a base class method. So understand the below in that context (i.e. we think the best fix involves using that information)

There are rules around widening that attempt to figure out what behavior you're trying expose in a method. If you wrote something like
```ts
    someMethod() {
       if (...) return 3; else return 6;
    }
```
we think that you want the signature `someMethod(): number`, not `someMethod(): 3 | 6` -- basically, TS shouldn't immediately expose the internals of the function without some kind of hint to do so.

We do infer `Date | false` as a return type when the return expressions are heterogeneous.

All these rules came from experimental data produced when we added literal types to the type system (in other words, we messed with them until we got the "best" set of breaking changes) and are based on our "best guess" about what the inferred return type should be. We already tweaked these rules to the best of our ability and any change here would move away from the local maximum.
